ao-tools/ao-telem/ao-telem.1
.\"
.\" Copyright © 2009 Keith Packard <keithp@keithp.com>
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\" General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License along
.\" with this program; if not, write to the Free Software Foundation, Inc.,
.\" 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
.\"
.\"
.TH AO-TELEM 1 "ao-telem" ""
.SH NAME
ao-telem \- Analyse a flight log (either telemetry or eeprom)
.SH SYNOPSIS
.B "ao-telem"
[\-s <summary-file>]
[\--summary=<summary-file>]
[\-d <detail-file>]
[\--detail=<detail-file>]
[\-r <raw-file>]
[\--raw=<raw-file>]
[\-p <plot-file>]
[\--plot=<plot-file>]
[\-g <gps-file]
[\--gps=<gps-file]
[\-k <kml-file]
[\--kml=<kml-file]
{flight.eeprom|flight.telem}
.SH DESCRIPTION
.I ao-telem
reads the specified flight log and produces several different kinds of
output.
.IP Summary
By default, summary information is shown on stdout. With the --summary
option, it can be redirected to a file.
.IP Detail
When requested with the --detail option, a filtered version of the
flight position, speed and acceleration are written to the specified
file.
.IP Raw
The --raw option writes the unfiltered, but converted acceleration
and height data to the specified file.
.IP Plot
The --plot option writes plots of height, speed and acceleration to
the specified file in .svg format
.IP GPS
The --gps option writes the recorded GPS data to the specified file in
three columns.
.IP KML
The --kml option writes the recorded GPS data to the specified file in
Keyhole Markup Language format, which can be displayed in Googleearth.
.SH AUTHOR
Keith Packard


ChangeLog
commit 93b8f40bb451c9ec152490d1f431ab18f8ecb7d1
Author: Bdale Garbee <bdale@gag.com>
Date:   Fri Aug 26 10:52:55 2011 -0600

    more release process doc updates

commit 9451ae5e4ea6ecfa512ba93197351682d4d043dc
Author: Bdale Garbee <bdale@gag.com>
Date:   Fri Aug 26 10:51:18 2011 -0600

    update release process docs

commit 01aee9ebe517ed657692e0a39a31ae0ddfb5b8b0
Author: Bdale Garbee <bdale@gag.com>
Date:   Fri Aug 26 10:39:45 2011 -0600

    update changelogs for Debian build

commit bc399d97a8424a5262f66a03e3bc403f793bb337
Author: Bdale Garbee <bdale@gag.com>
Date:   Fri Aug 26 10:38:44 2011 -0600

    roll back packaging changelog for rebuild

commit 13e6e799070a1469cbc2ff990379ee520b8f0e6a
Author: Bdale Garbee <bdale@gag.com>
Date:   Fri Aug 26 10:29:58 2011 -0600

    roll release notes version from 1.0 to 1.0.1

commit 1d15c841ad276127edae0345bd316c30731fffbf
Author: Bdale Garbee <bdale@gag.com>
Date:   Fri Aug 26 10:24:35 2011 -0600

    ignore generated log file

commit 58a74c3c3dcd01509224326cffeb67a3e809e4e6
Author: Bdale Garbee <bdale@gag.com>
Date:   Fri Aug 26 10:07:15 2011 -0600

    fix telemini firmware path name

commit 0a92eb2fa6b213533691288d8f99d72b80312983
Author: Bdale Garbee <bdale@gag.com>
Date:   Fri Aug 26 09:43:06 2011 -0600

    update changelogs for Debian build

commit 3cea033ec928c21b9f7b810898ed3c3cc536d2ce
Author: Bdale Garbee <bdale@gag.com>
Date:   Fri Aug 26 09:41:46 2011 -0600

    get ready for a 1.0.1 release

commit 674231773256bacd7acb4b5718c47412e47b813f
Merge: 08e3d54 3bfe8df
Author: Bdale Garbee <bdale@gag.com>
Date:   Fri Aug 26 09:37:01 2011 -0600

    Merge branch 'master' of ssh://git.gag.com/scm/git/fw/altos

commit 08e3d54bacf8b38a7c33e420a0bc19bcc7acedd4
Author: Bdale Garbee <bdale@gag.com>
Date:   Fri Aug 26 09:35:54 2011 -0600

    moving git-buildpackage config into .git/ since it is fairly specific to
    Bdale's build environment, and doesn't need to be in the source package

commit 3bfe8df44b575ca430ffaa051e20faa955a06c03
Author: Keith Packard <keithp@keithp.com>
Date:   Thu Aug 25 18:02:45 2011 -0700

    altos: Clear callsign on initial config load
    
    Before being written the first time, ao_config will get set to 0xff
    when the config storage is read. This leaves the tail of the callsign
    filled with invalid bytes. Zero the whole thing before loading the
    default callsign to make sure any extra bytes are set correctly.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 93ecaf7f18f8f4c15953c2e80dc9e1b20d04fdfa
Author: Bdale Garbee <bdale@gag.com>
Date:   Thu Aug 25 15:46:30 2011 -0600

    update TeleMini turnon script now that we've made a stable firmware release

commit c3314dae2d3df82e188daf6ba8520cce833592c6
Author: Bdale Garbee <bdale@gag.com>
Date:   Thu Aug 25 03:36:25 2011 -0600

    use multimaint-merge to make Debian changelogs less ugly

commit f9d87de0e2681c2ec5753f4fffef0e61a3a9e144
Author: Bdale Garbee <bdale@gag.com>
Date:   Thu Aug 25 02:22:53 2011 -0600

    update changelogs for Debian build

commit 41998645c04a2ee856dec272b40c2dc5913291d3
Author: Bdale Garbee <bdale@gag.com>
Date:   Thu Aug 25 02:16:17 2011 -0600

    add a postinst to remove sources.list.d fragment delivered by old private
    versions of this package, no longer needed once we have official packages

commit d65e4f6b0a8ba8f67c8a916f2e8be0ec5c75f47d
Author: Bdale Garbee <bdale@gag.com>
Date:   Thu Aug 25 02:05:49 2011 -0600

    really, I mean it, no longer deliver the sources.list fragment

commit 0579bd5de9e7686d227a16951dd520439ca5472b
Author: Bdale Garbee <bdale@gag.com>
Date:   Thu Aug 25 01:55:32 2011 -0600

    we need a main category in the desktop file

commit 3ec03792adebbeadc9c0d4985c6f51877e8ab969
Author: Bdale Garbee <bdale@gag.com>
Date:   Thu Aug 25 01:48:11 2011 -0600

    update desktop file for consistency with package section

commit 0b1d5b678f0e36fbe09d4d3babffb77f857f098d
Author: Bdale Garbee <bdale@gag.com>
Date:   Thu Aug 25 01:37:31 2011 -0600

    rewind packaging changelog to last tagged version

commit f0891fc2e4fa4fd6dd5c2bfc74e4514de986a295
Author: Bdale Garbee <bdale@gag.com>
Date:   Thu Aug 25 01:35:18 2011 -0600

    ao-view is no longer included in the altos package, so have the old-style
    Debian menu entry point to altosui instead

commit a7ee1049519bf46b3864666999249876009fc3bf
Author: Bdale Garbee <bdale@gag.com>
Date:   Thu Aug 25 01:35:06 2011 -0600

    changes in preparation for upload to Debian

commit 2992184f7a032c95e22023080fbc26443ed786f5
Author: Bdale Garbee <bdale@gag.com>
Date:   Thu Aug 25 01:30:36 2011 -0600

    don't deliver sources.list fragment in official Debian packages

commit 3fc92a529dae410a41038cce38c6a44f5e58bd1f
Author: Bdale Garbee <bdale@gag.com>
Date:   Thu Aug 25 01:11:47 2011 -0600

    prepare to release

commit 73abe19acf709c00f5352ec12e8cd6edae1d1963
Merge: 1bd781d 5158493
Author: Bdale Garbee <bdale@gag.com>
Date:   Thu Aug 25 00:34:49 2011 -0600

    Merge branch 'master' of ssh://git.gag.com/scm/git/fw/altos
    
    Conflicts:
    	doc/altusmetrum.xsl

commit 1bd781da934c738e0c9294197c7eb622b0710a9a
Author: Bdale Garbee <bdale@gag.com>
Date:   Thu Aug 25 00:32:47 2011 -0600

    more tweaks

commit 5158493c8df527e7527057c719c75248609eb3dc
Author: Keith Packard <keithp@keithp.com>
Date:   Wed Aug 24 23:21:02 2011 -0700

    doc: Remove duplicate documentation about max flight log
    
    This was described in detail in both the System Operation and AltosUI
    chapters. Remove the duplicate from the AltosUI chapter.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit e268798dc260311f5f0167909481b41c9d27fc1c
Merge: 458f816 242344d
Author: Keith Packard <keithp@keithp.com>
Date:   Wed Aug 24 23:06:44 2011 -0700

    Merge remote-tracking branch 'origin/master'

commit 458f816ad23fd6784757b13b244057d4be64260e
Author: Keith Packard <keithp@keithp.com>
Date:   Wed Aug 24 23:06:01 2011 -0700


ao-tools/ao-telem/ao-telem.c
/*
 * Copyright © 2011 Keith Packard <keithp@keithp.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; version 2 of the License.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
 */

#define _GNU_SOURCE
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <getopt.h>
#include "cc.h"

static const struct option options[] = {
	{ 0, 0, 0, 0},
};

static void usage(char *program)
{
	fprintf(stderr, "usage: %s\n"
		"\t{flight-log} ...\n", program);
	exit(1);
}

int
main (int argc, char **argv)
{
	char	line[80];
	int c, i, ret;
	char *s;
	FILE *file;
	int serial;
	while ((c = getopt_long(argc, argv, "", options, NULL)) != -1) {
		switch (c) {
		default:
			usage(argv[0]);
			break;
		}
	}
	for (i = optind; i < argc; i++) {
		file = fopen(argv[i], "r");
		if (!file) {
			perror(argv[i]);
			ret++;
			continue;
		}
		s = strstr(argv[i], "-serial-");
		if (s)
			serial = atoi(s + 8);
		else
			serial = 0;
		while (fgets(line, sizeof (line), file)) {
			union ao_telemetry_all telem;
			char call[AO_MAX_CALLSIGN+1];
			char version[AO_MAX_VERSION+1];

			if (cc_telemetry_parse(line, &telem)) {
				printf ("serial %5d tick %5d type %3d ",
					telem.generic.serial, telem.generic.tick, telem.generic.type);
				switch (telem.generic.type) {
				case AO_TELEMETRY_SENSOR_TELEMETRUM:
				case AO_TELEMETRY_SENSOR_TELEMINI:
				case AO_TELEMETRY_SENSOR_TELENANO:
					printf ("state %1d accel %5d pres %5d ",
						telem.sensor.state, telem.sensor.accel, telem.sensor.pres);
					printf ("accel %5d speed %5d height %5d ",
						telem.sensor.acceleration,
						telem.sensor.speed,
						telem.sensor.height);
					printf ("ground_pres %5d ground_accel %5d accel_plus %5d accel_minus %5d\n",
						telem.sensor.ground_pres,
						telem.sensor.ground_accel,
						telem.sensor.accel_plus_g,
						telem.sensor.accel_minus_g);
					break;
				case AO_TELEMETRY_CONFIGURATION:
					memcpy(call, telem.configuration.callsign, AO_MAX_CALLSIGN);
					memcpy(version, telem.configuration.version, AO_MAX_VERSION);
					call[AO_MAX_CALLSIGN] = '\0';
					version[AO_MAX_CALLSIGN] = '\0';
					printf ("device %3d flight %5d config %3d.%03d delay %2d main %4d",
						telem.configuration.device,
						telem.configuration.flight,
						telem.configuration.config_major,
						telem.configuration.config_minor,
						telem.configuration.apogee_delay,
						telem.configuration.main_deploy,
						telem.configuration.flight_log_max);
					printf (" call %8s version %8s\n", call, version);
					break;
				default:
					printf("\n");
				}
			}
		}
		fclose (file);

	}
	return ret;
}
